midi/midi/encode.py

```diff
@@ -4,227 +4,245 @@
-    midi_file.tracks.append(get_tempo_meta_messages(tempos, acc_factor))
+    accuracy = float(960 / grid_accuracy)  # = ticks_per_beat / (notated_32nd_notes_per_beat * grid_accuracy / 32)
+    if event_lengths is not None:
+        tempos = get_tempo_array_from_tempo_sequences(tempos, event_lengths)
+    midi_file.tracks.append(get_tempo_meta_messages(tempos, accuracy))
 
-    for i in range(input_array.shape[0]):
-        midi_file.tracks.append(get_note_messages_from_2d_array(input_array[i], i, acc_factor))
-    midi_file.save(f'../../outputs/{output_filename}.mid')  # TODO: parametrization
+    return tempos, accuracy, midi_file
 
 
-def generate_file_from_midi_features(input_array, output_filename, accuracy):   # TODO: check if > 0
+def get_messages_from_standard_2d_input(data, track_channel, tempos, accuracy, join_notes, use_sequences,
+                                        use_velocities, event_lengths=None):
     """
-    translates a music_21.py output matrix of MIDI features into a MIDI file
-
-    :param input_array:
-    :param output_filename:
-    :param accuracy:
-    :return:
+    translates a two-dimensional single-track array into a MidiTrack
     """
-    if input_array.ndim != 2:
-        raise TypeError('input array must have exactly 2 dimensions')
+    if data.shape[0] != len(tempos):
+        raise IndexError('length of tempos array and input array\'s time dimension must be equal')
 
-    tempos = []
-    array = []
+    if use_sequences and event_lengths is None:
+        raise ValueError('no argument \'event_lengths\' for \'use_sequences\' mode provided')
 
-    for feature_index in range(input_array.shape[1] - 2):
-        track = []
-        for event_index in range(input_array.shape[0]):
-            notes = [False] * 128
-            feature = round(input_array[event_index][feature_index] * 128 - 1)
-            if feature != 0:
-                notes[feature] = True
+    if use_sequences:
+        events = data.tolist()
+    else:
+        events, event_lengths = get_sequences_from_array(data)
 
-            length = round(1 / input_array[event_index][-2])
-            track.extend([notes] * length)
+    events = get_tuples_from_sequences(events)
+    track = get_messages_from_tuples(events, track_channel, event_lengths, accuracy, join_notes, not use_velocities)
 
-        array.append(track)
+    return track
 
-    for event_index in range(input_array.shape[0]):
-        length = round(1 / input_array[event_index][-2])
-        tempo = round(1000 / input_array[event_index][-1])
-        tempos.extend([tempo] * length)
-    tempos.extend([round(1000 / input_array[-1][-1])])
 
-    array = np.asarray(array)
-    generate_file_from_3d_array(array, tempos, output_filename, accuracy)
+def get_file_from_standard_features(data, tempos, output_path, join_notes, use_sequences, use_velocities,
+                                    event_lengths=None, grid_accuracy=GRID_ACCURACY):  # TODO: create if doesnt exist
```

Comment:
Bump todo?

Reply:
`output_path - create directory if doesnt exist` will be added in future commits